src/Corax/Utils/Bm25Relevance.cs

```diff
@@ -0,0 +1,191 @@
+    private int _currentId;
+    private readonly float _idf;
+    private PostingList.Iterator _setIterator;
+
+
+    public readonly bool IsStored;
+    public readonly bool IsInitialized;
+    private bool _isDisposed;
+    private readonly int _bufferCapacity;
+    private Span<long> Matches => new(_matchBuffer, _currentId);
+    private Span<short> Scores => new(_scoreBuffer, _currentId);
+
+    private Bm25Relevance(IndexSearcher indexSearcher, long termFrequency, ByteStringContext context, int numberOfDocuments, double termRatioToWholeCollection,
+        delegate*<ref Bm25Relevance, Span<long>, Span<float>, float, void> dynamicalScoreFunc)
+    {
+        IsInitialized = true;
+        _termRatioToWholeCollection = (float)termRatioToWholeCollection;
+        _numberOfDocuments = numberOfDocuments;
+        IsStored = MaximumDocumentCapacity > numberOfDocuments;
+
+
+        if (IsStored == false && dynamicalScoreFunc != null)
+        {
+            _scoreFunc = dynamicalScoreFunc;
+            _processFunc = &DecodeAndDiscard;
+            _bufferCapacity = MaximumDocumentCapacity;
+            _currentId = MaximumDocumentCapacity;
+        }
+        else
+        {
+            _processFunc = &DecodeAndSave;
+            _scoreFunc = &CalculateScoreFromMemory;
+            _bufferCapacity = numberOfDocuments;
+            _currentId = 0;
+        }
+
+        context.Allocate(_bufferCapacity * (sizeof(long) + sizeof(short)), out var buffer);
+        _matchBuffer = (long*)buffer.Ptr;
+        _scoreBuffer = (short*)(buffer.Ptr + _bufferCapacity * sizeof(long));
+
+        _idf = ComputeIdf(indexSearcher, termFrequency);
+    }
+
+    private static float ComputeIdf(IndexSearcher indexSearcher, long termFrequency)
+    {
+        var m = indexSearcher.NumberOfEntries - termFrequency + 0.5D;
+        var d = termFrequency + 0.5D;
+
+        return (float)Math.Log((m / d) + 1);
+    }
+
+    public void Score(Span<long> matches, Span<float> scores, float boostFactor)
+    {
+        if (_isDisposed)
+            ThrowAlreadyDisposed();
+
+        _scoreFunc(ref this, matches, scores, boostFactor);
+    }
+
+    //Requirements: ids are sorted!!!
+    private static void CalculateScoreFromMemory(ref Bm25Relevance bm25, Span<long> matches, Span<float> scores, float boostFactor)
+    {
+        //The score will be almost 0. Don't care.
+        if (bm25._idf.AlmostEquals(0f))
+            return;
+
+        var innerItems = bm25.Matches;
+        var frequencies = bm25.Scores;
+
+        for (int idX = 0; idX < matches.Length; ++idX)
+        {
+            var entryId = matches[idX];
+            var idOfInner = innerItems.BinarySearch(entryId);
+
+            if (idOfInner < 0)
+                continue;
+
+            var weight = frequencies[idOfInner] / ((1 - BFactor) + BFactor * bm25._termRatioToWholeCollection);
+            scores[idX] += bm25._idf * weight * boostFactor / (K1 + weight);
+        }
```

Comment:
Copy or summarize the relevant parts of the paragraphs from the document. And provide location markers to where this is explained in the paper.

Reply:
done